automatic-config-updates/README.md

```diff
@@ -0,0 +1,67 @@
+
+A disadvantage of 1) is that it requires user action (in the form of a PR).
+If a project that has Packit enabled doesn't have any PR for a few months, during which we change the config and remove support for the old format, things like triggering `propose-downstream` won't work because the config will be invalid (because there was no PR where we could update it) and unparsable.
+With this in mind, I consider 2) to be a must-have, while 1) to be a nice-to-have.
+
+On the other hand, 1) has the advantage that it can fix the config of the user before it is merged to main if they try to use a deprecated feature, which 2) can't really handle (the config would be merged and then we would create a PR with the correction, so a correct config will end up on the main branch in the end).
+
+## Implementation approach
+
+Regardless of which of the following approaches ends up being used, the process of updating the config will always be the following:
+
+1. Get the config, determine if transformations need to be done
+2. Pull the repository
+3. Make transformations to the config
+4. Create a pull request (or add a commit to a PR) with the transformations
+
+Let's take a closer look at the more "problematic" parts of this pipeline.
+
+### Getting the config
+
+There seem to be three main approaches to getting the config:
+
+1. In case we do the modifications by adding commits to PRs, the config should come from the PR.
+2. Get the list of projects by querying `/api/projects` and then find configs in them.
+3. Add a new database table mapping project to its config content on main.
+
+Let's focus on comparing 2) and 3) since those are relevant to the periodic updates.
+Getting the list of projects with Packit enabled is straight-forward. However, we would have to do quite a lot of queries to Github to get all the configs (we could even run into Github's request quota).
+Moreover, with this approach, we can't store any additional information about the config, which could perhaps be useful.
+
+On the other hand, 3) would spread the requests to Github to get the current config across time. We would basically have to create an implicit action that triggers on merge to main/master and fetches the current config to the database. Hence, we wouldn't be making a lot of requests to Github at the same time, but rather over time.
+This approach also facilitates storing additional data about the configs, for example which transformations have already been applied (see the following section).
+Such additional information could allow us to detect that there are no transformations to do without even checking the config (resulting in faster overall updates).
+The main downside of storing the configuration in the database are the space requirements.
+However, this should hopefully not be a problem, since the config is usually not that long (but it is still something to keep in mind).
```

Comment:
We could have a dedicated storage/database for this. Ideally something that allows easy searching in json structure.

Reply:
Sounds like a task for document NoSql database (Mongo?)

Reply:
How would a separate/dedicated storage/database help with space requirements?
It'd just take the space somewhere else, wouldn't it?